File: Assets/Scripts/Base/Pool/ClassObjectPool.cs
/**
 * 类对象池
 */

using System.Collections.Generic;

namespace Base.Pool
{
    public class ClassObjectPool<T> where T : class, new()
    {
        /// <summary>
        /// The pool
        /// </summary>
        protected Stack<T> mPool = new Stack<T>();
        /// <summary>
        /// 最大对象个数，-1表示不限个数
        /// </summary>
        protected int mMaxCount = 0;
        /// <summary>
        /// 没有回收的对象个数
        /// </summary>
        protected int mNoRecycleCount = 0;

        /// <summary>
        /// 初始化默认为不限个数
        /// </summary>
        public ClassObjectPool(int maxCount = -1)
        {
            mMaxCount = maxCount;
        }

        /// <summary>
        /// 分配类对象
        /// </summary>
        public T Spawn()
        {
            mNoRecycleCount++;
            if (mPool.Count > 0)
            {
                T obj = mPool.Pop();
                if (obj == null)
                {
                    obj = new T();
                }
                return obj;
            }
            else
            {
                T obj = new T();
                return obj;
            }
        }

        /// <summary>
        /// 回收对象
        /// </summary>
        public bool Recycle(T obj)
        {
            if (obj == null) return false;
            mNoRecycleCount--;

            if (mPool.Count >= mMaxCount && mMaxCount > 0)
            {
                obj = null;
                return false;
            }

            mPool.Push(obj);
            return true;
        }
    }
}

File: Assets/LuaScripts/Base/Pool/ClassObjectPool.lua
--[[
    @desc:类对象池
]]

local ClassObjectPool = Class("ClassObjectPool")

-- maxCount -1表示不限个数
local function Ctor(self, class, maxCount)
    self.mPool = Stack:New()
    self.mClass = require(class)
    self.mMaxCount = maxCount or -1
end

-- 分配对象
local function Spawn(self)
    if self.mPool:Count() > 0 then
        print("现有")
        return self.mPool:Pop()
    else
        print("创新")
        return self.mClass:New()
    end
end

-- 回收对象
local function Recycle(self, obj)
    if obj == nil then return end
    if self.mMaxCount > 0 and self.mPool:Count() >= self.mMaxCount then
        obj = nil
        return
    end
    self.mPool:Push(obj)
end

-- 清空对象池
local function Clear(self)
    self.mPool:Clear()
end

ClassObjectPool.Ctor = Ctor
ClassObjectPool.Spawn = Spawn
ClassObjectPool.Recycle = Recycle
ClassObjectPool.Clear = Clear

return ClassObjectPool